aboozaid/react-native-wifi-hotspot
# Refactor to use new `startLocalOnlyHotspot` logic

## Purpose

This PR refactors the library to use the new [`startLocalOnlyHotspot`](https://developer.android.com/reference/android/net/wifi/WifiManager#startLocalOnlyHotspot(android.net.wifi.WifiManager.LocalOnlyHotspotCallback,%20android.os.Handler)) method that is available in Android Oreo and above. 


## Changes

The bulk of the file changes are due to the removal of the old `Example` React Native project, and the creation of a new React Native Project, `TestApp`. Creating an entirely new project was easier than trying to make all of the patches necessary for the old project.

### Local only hotspot

This is the core of the changes to the library. As of Android APK 28+, developers can only programmatically create a hotspot that does not have internet access.

### Removal of custom configuration options
`startLocalOnlyHotspot` has access to the [`SoftApConfiguation`](https://developer.android.com/reference/android/net/wifi/SoftApConfiguration) for the hotspot but does not have the ability to pass custom configuration to the AP.

## Testing 

To test, we will run the `TestApp` project using a local version of the project instead of the one pushed to NPM. 

I ran into some issues with resolving the correct `node_modules` when trying to link the two together in the same directory, so these instructions create a copy of `TestApp`.

### Setup

1. `cp -r TestApp ~/Desktop && cd ~/Desktop/TestApp` (`~/Desktop` can be replaced with any directory)
2. Open `package.json` and remove the `react-native-wifi-hotspot` hotspot dependency.
3. `yarn && yarn add file:/absolute/path/to/react-native-wifi-hotspot`
4. Follow the [setup instructions](https://github.com/Patrick-Erichsen/react-native-wifi-hotspot/tree/pe-local-only-hotspot#setup), skipping the first install step for npm/yarn.
5. `react-native run-android`

### Validation
- Press the "Start" button and confirm `Hotspot connection started` is displayed
- Press the "Show Peers" button and confirm a list of peers is displayed
- Press the "Show config" button and confirm the config is displayed
- Press the "Close" button and confirm `Hotspot connection closed` is displayed

## Future work

Based on these updates, there are a few future items of work that should likely be implemented:

- `startLocalOnlyHotspot` provides a callback method for when the hotspot is actually torn down that is not implemented in this PR. The `close` callback that is provided is only invoked when the apps connected to the underlying hotspot is closed, but since other apps could be using the hotspot, it is not necessarily torn down.
- Fixing the bug that requires a path update to the package in `Settings.gradle` (described in step 3 of setup)
